### PR TITLE
Correction of U-Value Calculation with Film Coefficient

### LINKED ISSUE #9470: Air film R-value for semi-exterior surfaces not used
Issue overview
--------------
Section A9.4.1.2 in ASHRAE Standard 90.1-2019 states that "semi-exterior surfaces are protected surfaces that face attics, crawlspaces and parking garages with natural or mechanical ventilation".

EnergyPlus currently uses the prescribed semi-exterior surface air film R-value as the default in the following `switch` statement used to calculate the U-value of opaque interior surfaces. This value ends up not being used for what it is intended to represent. For example, the U-value with film for the small office prototype building model should be calculated using this value but is currently calculated using the R-value for  "interior horizontal surfaces, heat flow up" and "interior horizontal surfaces, heat flow down".

https://github.com/NREL/EnergyPlus/blob/317b0c131df2eacad09c2224236658ab45b3afa4/src/EnergyPlus/DataHeatBalance.cc#L1321-L1340

### Details
Some additional details for this issue (if relevant):
 - Version of EnergyPlus: 22.1

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Defect file added (list location of defect file here)
 - [ ] Ticket added to Pivotal for defect (development team task)
 - [ ] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

### PR DESCRIPTION
**Pull request by RKStrand**
Pull request overview
---------------------
 - Fixes #9470 
 - This pull request is to fix an issue that was noticed in the code where the film coefficient for "semi-exterior" surfaces was being used for other things including a default value improperly.  There were also some cases missing from the existing logic due to the addition of new surfaces and a translation of this code from F90 to c++.  All of this has been fixed, a new unit test was created to exercise this modified code, and documentation edits were made to clean some things up and explain how this impacts the results.  The value being calculated only impacts the EIO file since it is a reporting value that is not used to calculate anything in the code.  No diffs in the simulation results were noted other than the changes to this report value in the EIO.

**NOTE: ENHANCEMENTS MUST FOLLOW A SUBMISSION PROCESS INCLUDING A FEATURE PROPOSAL AND DESIGN DOCUMENT PRIOR TO SUBMITTING CODE**

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [X] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [X] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [X] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [x] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [x] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [x] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

File: src/EnergyPlus/DataHeatBalance.cc
Comment (amirroth) [inline src/EnergyPlus/DataHeatBalance.cc]: If you are only trying to assign a constant value based on an enumeration, a smaller/faster idiom is:

```cpp
static constexpr std::array<Real64, static_cast<int>DataSurfaces::SurfaceClass::Num> outsideFilmCoefs = { 
   0.1197584, // Wall
   0.1620212, // Floor
   0.1074271, // Roof 
   ... 
};

outsideFilm = outsideFilmCoefs[static_cast<int>(state.dataSurface->Surface(thisSurface.ExtBoundCond).Class)];
```

That converts an entire `switch` into a single lookup. There are a few examples of this in the code already, e.g., in `ConvectionCoefficients`.

BTW Given that we have `Invalid` is `None` a valid `SurfaceClass` value?

Comment (amirroth) [reply src/EnergyPlus/DataHeatBalance.cc:1328]: That converts an entire switch/case into a single lookup.

Comment (RKStrand) [reply src/EnergyPlus/DataHeatBalance.cc]: I have pushed a new commit that changes this as you suggested.

As for Invalid vs. None, I don't know that "None" has any specific reason to be there.  A search of SurfaceClass::None in the code only results in four occurrences.  I'll see if I can get rid of None without too much trouble and make another commit if I don't run into any complications.

Comment (RKStrand) [reply src/EnergyPlus/DataHeatBalance.cc]: Ok, I've tried this and while None is not really used for anything the integer evaluation of SurfaceClass is used as indices for other arrays.  It skips the "None" which is evaluating to 0.  Getting rid of None means that Wall becomes zero which leads to index errors.  So, it looks like there is a "use" for None though maybe at some point things could be rewritten to get rid of that case.  For now, I plan on leaving None in and move forward with the other changes unless someone has an issue with that.

Comment (amirroth) [reply src/EnergyPlus/DataHeatBalance.cc]: This is an artefact of Fortran 1-based arrays and the use of 0 to mean "not found" or "invalid". In the C/C++ world, 0 is valid of course and -1 is used for "invalid" or "not found". 

It may be worth taking a quick stab to see if the arrays indexed by `SurfaceClass` (and which are causing the index errors) can be switched over from `Array1D` to `std::array` without requiring knock-on changes.  If so, great. If not, 📌 . If too tired, also 📌 .

Comment (RKStrand) [reply src/EnergyPlus/DataHeatBalance.cc]: Let's 📌 this for now.  I'm wondering if at some point someone will need to look beyond just this one and find other items like this.  Hopefully the other changes here are solid and ci continues to come back clean.

File: doc/output-details-and-examples/src/output-files/eplusout-eio.tex
Comment (rraustad) [inline doc/output-details-and-examples/src/output-files/eplusout-eio.tex:1828]: OutputDetailsAndExamples.pdf does compile and reflects changes shown above.

Very picky comment, this should be a semi-colon (no need to correct unless other commits are pending):

image

These look correct (except there is no underscore for Line 2 and Ex Line 2's, again a picky comment). Probably been a while since we have looked at these:

image

    ! <Zone Surfaces>,Zone Name,# Surfaces
    ! <Shading Surfaces>,Number of Shading Surfaces,# Surfaces
    ! <Shading Surface>,Surface Name,Surface Class,Base Surface,Heat Transfer Algorithm,Transmittance Schedule,Min Schedule Value,Max Schedule Value,Solar Diffusing,Area (Net) {m2},Area (Gross) {m2},Area (Sunlit Calc) {m2},Azimuth {deg},Tilt {deg},~Width {m},~Height {m},Reveal {m},ExtBoundCondition,ExtConvCoeffCalc,IntConvCoeffCalc,SunExposure,WindExposure,ViewFactorToGround,ViewFactorToSky,ViewFactorToGround-IR,ViewFactorToSky-IR,#Sides
    ! <Frame/Divider Surface>,Surface Name,Surface Class,Base Surface,Heat Transfer Algorithm,Construction,Nominal U (w/o film coefs) {W/m2-K},Nominal U (with film coefs) {W/m2-K},Solar Diffusing,Area (Net) {m2},Area (Gross) {m2},Area (Sunlit Calc) {m2},Azimuth {deg},Tilt {deg},~Width {m},~Height {m},Reveal {m}
    Shading Surfaces,Number of Shading Surfaces,4
    Shading Surface,MAIN SOUTH OVERHANG,Shading,FRONT-1,None,SHADETRANSSCH,0.00,0.00, ,25.74,25.74,25.74,210.00,180.00,19.80,1.30,,,,,,,,,,,4
    Zone Surfaces,PLENUM-1,10
    HeatTransfer Surface,WALL-1PF,Wall,,CTF - ConductionTransferFunction,WALL-1,0.408,0.384,,18.30,18.30,18.30,210.00,90.00,30.50,0.60,0.00,ExternalEnvironment,ASHRAESimple,ASHRAESimple,SunExposed,WindExposed,0.50,0.50,0.52,0.48,4

File: src/EnergyPlus/DataHeatBalance.cc
Comment (rraustad) [inline src/EnergyPlus/DataHeatBalance.cc:1354]: This change looks good.

File: tst/EnergyPlus/unit/DataHeatBalance.unit.cc
Comment (rraustad) [inline tst/EnergyPlus/unit/DataHeatBalance.unit.cc:1364]: Unit test looks good.

**Comment by rraustad** [comment]
Other than very picky comments this change looks good.

**Comment by rraustad** [comment]
@RKStrand can you comment on the diffs please. There are expected diffs but to complete this a comment on diffs is warranted.

**Comment by RKStrand** [comment]
@rraustad The only diffs should be in the EIO output because the calculations impacted are for reporting purposes only.  The values produced before the fix were not correct because the wrong coefficient.  The new values are cleared and more correct.  Also, the new docs acknowledge that the calculations are a guess since we don't have all the information to make this calculation, but the calculation is pretty simple to do outside of E+.  Does that help or do you need more information?  Or are there diffs in other files (though this might need to brought up to develop).

**Comment by rraustad** [comment]
This branch is your 3 commits ahead of develop. The first diff I looked at was 1ZoneWith14ControlledHeat-CoolPanels. Are these differences expected?

image

**Comment by RKStrand** [comment]
Wow--I found a very subtle error in one of the coefficients (two numbers switched around).  Hopefully this fixes things.  Some minor differences in the EIO are expected.  The ones in the panels in the file you mention above are very exaggerated because any small difference in the film coefficients will result in very large changes in the U-value because the panel itself has nearly no resistance (metal sheet basically).  Just pushed that commit.  Hopefully this will the number and/or severity of the diffs.

File: src/EnergyPlus/DataHeatBalance.cc
Comment (rraustad) [inline src/EnergyPlus/DataHeatBalance.cc:1340]: Where did this semi-exterior coefficient (0.0810106) get moved to? or was it never used?

File: src/EnergyPlus/DataHeatBalance.cc
Comment (rraustad) [inline src/EnergyPlus/DataHeatBalance.cc]: These coefficients are numerically correct now and same as before. The array lines up with enum class SurfaceClass in DataSurfaces.hh

File: src/EnergyPlus/DataHeatBalance.cc
Comment (RKStrand) [reply src/EnergyPlus/DataHeatBalance.cc:1340]: It should never have been used (was used in error) and was removed from the code.  There's nothing in the input that would indicate definitely a semi-exterior surface.  It was used as a default (incorrectly) before.

File: src/EnergyPlus/DataHeatBalance.cc
Comment (RKStrand) [reply src/EnergyPlus/DataHeatBalance.cc]: For some reason, even though your comment was positive, it jogged my memory that the coefficient that had to be fixed shows up three other times in this same statement. 🤦‍♂️  Just committed that fix.  Trying to juggle too many branches at once...

**Comment by rraustad** [comment]
Review complete. Waiting on CI results to check for diffs and performance impact.

File: src/EnergyPlus/DataHeatBalance.cc
Comment (rraustad) [reply src/EnergyPlus/DataHeatBalance.cc]: Wow, I looked right at those, and yes they are different in the last 2 decimal places. Reviewing code sure can be difficult at times. OK, so now my comment is accuracte. Thanks for catching those.

**Comment by rraustad** [comment]
~~It looks to me as if the html diffs are justified.~~ There are no eio diffs, just table diffs. Running the first diff file in develop and this branch shows where the difference occurs. @RKStrand please acknowledge this diff is because of a correction to the code.

image

**Comment by rraustad** [comment]
@RKStrand should the outside surface coefficient of a door (panel002) in the roof surface have an outside film coefficient of 0 ?

**Comment by rraustad** [comment]
Looking for a reason why the outside film coefficient should be 0, I see this as the doors base surface which uses other side coefficients. So does this door (in a roof) use the base surfaces OtherSideCoefficients or it's own? And if the doors Outside Boundary Condition Object is blank, what outside film coefficient should be used?

    FenestrationSurface:Detailed,
      panel002,                !- Name
      Door,                    !- Surface Type
      PanelConst,              !- Construction Name
      Zn001:Roof001,           !- Building Surface Name
      ,                        !- Outside Boundary Condition Object

    BuildingSurface:Detailed,
      Zn001:Roof001,           !- Name
      Roof,                    !- Surface Type
      ROOF31,                  !- Construction Name
      ZONE ONE,                !- Zone Name
      ,                        !- Space Name
      OtherSideCoefficients,   !- Outside Boundary Condition
      Zn001:Roof001:OSC,       !- Outside Boundary Condition Object

    SurfaceProperty:OtherSideCoefficients,
      Zn001:Roof001:OSC,       !- Name
      0,                       !- Combined Convective/Radiative Film Coefficient {W/m2-K}
      0,                       !- Constant Temperature {C}
      0.95,                    !- Constant Temperature Coefficient
      ,                        !- External Dry-Bulb Temperature Coefficient
      ,                        !- Ground Temperature Coefficient

**Comment by RKStrand** [comment]
I've tracked this down and have a reason why it is doing what it is doing.  The surface uses Other Side Coefficients.  As a result, the outside environment is not exterior but other side coefficients.  In this case, I don't think that develop is right--it's not necessarily a semi-exterior surface.  This branch chose to set those to zero because other side coefficients kinda mess with the heat balance equations.  In other words, it's a completely different situation and these U-values are somewhat useless.  So, that is why you are seeing the zero exterior film coefficient here.  Is that what we want?  I guess that's a matter of debate.  If you have strong feelings, let me know and I'll implement it.

**Comment by RKStrand** [comment]
I would think as a subsurface the window would inherit the OSC from the base surface (Roof).  Based on the new algorithm, it must have because that is why the exterior film is 0.

**Comment by RKStrand** [comment]
If it's any help, the film coefficient is actually listed as zero in this example. 😂  I did actually put some language in the docs to let people know that it can't identify semi-exterior.  I could add that when using other side coefficients that the results need to be adjusted by the user.  Thoughts?

**Comment by rraustad** [comment]
The only strong feelings I have is getting this right before it's merged. I don't know what the correct fix is. It seems to me that a surface exposed to outdoors, or more specifically other side coefficients, should have a non-zero film coefficient. In this case the user specified a 0 film coefficient for this base surface. The other side coefficients film coefficient for this sub-surface might be 0 (I'm not sure because this FenestrationSurface:Detailed object does not specify an Outside Boundary Condition Object, the base surface does), but the code as it stands now is not looking at that coefficient, it just uses a hard 0. Should the case for other side coefficients use the value shown here? It may be that ... If the surface OtherSideCoefficients is not blank use that coefficient, if the surface OtherSideCoefficients is blank, use the base surface coefficient. Does that sound reasonable?

    SurfaceProperty:OtherSideCoefficients,
      Zn001:Roof001:OSC,       !- Name
      0,                       !- Combined Convective/Radiative Film Coefficient {W/m2-K}

**Comment by rraustad** [comment]
Let me be clear what I am talking about. This code is not using the OSC value, it just uses 0.

image

**Comment by rraustad** [comment]
@RKStrand I am thinking something like this. Since the 1ZoneWith14ControlledHeat-CoolPanels example file uses 0 as the film coefficient the ExtBoundCond == OtherSideCoefNoCalcExt. So in that case the outsideFilm coefficient should be 0. If I change the film coefficient to 0.08 then ExtBoundCon == OtherSideCoefCalcExt. So should this calculation, in that case, use outsideFilm = state.dataSurface->OSC(thisSurface.OSCPtr).SurfFilmCoef ?

    switch (thisSurface.ExtBoundCond) {
    case DataSurfaces::ExternalEnvironment: { // ExtBoundCond = 0
        outsideFilm = 0.0299387;              // All exterior conditions
    } break;
    case DataSurfaces::OtherSideCoefCalcExt: {
        outsideFilm = state.dataSurface->OSC(thisSurface.OSCPtr).SurfFilmCoef; <-- use the non-zero film coefficient
    } break;
    case DataSurfaces::Ground:
    case DataSurfaces::OtherSideCoefNoCalcExt:
    case DataSurfaces::OtherSideCondModeledExt:
    case DataSurfaces::GroundFCfactorMethod:
    case DataSurfaces::KivaFoundation: { // All these cases have a negative ExtBoundCond so don't use film coefficients
        outsideFilm = 0.0;

1ZoneWith14ControlledHeat-CoolPanels modified to use a non-zero film coefficient:

    SurfaceProperty:OtherSideCoefficients,
      Zn001:Roof001:OSC,       !- Name
      0.08,                       !- Combined Convective/Radiative Film Coefficient {W/m2-K}
      0,                       !- Constant Temperature {C}
      0.95,                    !- Constant Temperature Coefficient

image

**Comment by RKStrand** [comment]
@rraustad I am definitely open to something like what you are proposing. I will try to implement something like it later today and make another commit. Thanks for burning the midnight and weekend oil for all these reviews!

**Comment by RKStrand** [comment]
@rraustad Ok, I just committed this change along with an additional unit test to cover this new case and documentation improvements including the ones you mentioned above and a few others.  Hoping this passes the iteration tests!

**Review by rraustad — APPROVED** [review]
@Myoldmopar @mjwitte I am good with these changes. However, this is important enough to have more than 1 reviewer looking at these changes.

File: src/EnergyPlus/DataHeatBalance.cc
Comment (mjwitte) [reply src/EnergyPlus/DataHeatBalance.cc]: Using the array is nice and tidy, but it comes with the risk of misalignment if the surface class list is rearranged or added to in the future. A comment by the enum delcaration referring to this array is in order to alert future developers. I haven't examined the unit test carefully, but does it cover every enum case and compare against a hard value in the unit test so that it will fail if there's any misalignment?

Comment (amirroth) [reply src/EnergyPlus/DataHeatBalance.cc]: try static_assert.

**Comment by RKStrand** [comment]
Added in next commit.

On Sep 13, 2022, at 3:21 PM, Michael J. Witte ***@***.******@***.***>> wrote:



@mjwitte commented on this pull request.

________________________________

In src/EnergyPlus/DataHeatBalance.cc<https://urldefense.com/v3/__https://github.com/NREL/EnergyPlus/pull/9639*discussion_r970049085__;Iw!!DZ3fjg!5GpLYrzXHMU4mSymaeQZZGz0ArSkwrFRAOabu8X7V1JGlpyifqHKd4OBnDwRSxG7U9gvbcHW229n-vUUDwDzx_TX7R0$>:

> +        0.0,       // None
+        0.1197548, // Wall
+        0.1620212, // Floor
+        0.1074271, // Roof
+        0.0,       // IntMass
+        0.0,       // Detached_B
+        0.0,       // Detached_F
+        0.1197584, // Window
+        0.1197584, // GlassDoor
+        0.1197584, // Door
+        0.0,       // Shading
+        0.0,       // Overhang
+        0.0,       // Fin
+        0.0,       // TDD_Dome
+        0.0        // TDD_Diffuser
+    };


Using the array is nice and tidy, but it comes with the risk of misalignment if the surface class list is rearranged or added to in the future. A comment by the enum delcaration referring to this array is in order to alert future developers. I haven't examined the unit test carefully, but does it cover every enum case and compare against a hard value in the unit test so that it will fail if there's any misalignment?

**Comment by RKStrand** [comment]
Not sure why this isn’t letting me comment on GitHub, but here’s an email response.  Still a bit new to static_assert.  I tried a variety of versions of it, but I couldn’t get a direct static_assert to work for some reason (user error no doubt).  I put a work around in the unit test, though it’s a bit clunky.

On Sep 13, 2022, at 5:21 PM, amirroth ***@***.******@***.***>> wrote:



@amirroth commented on this pull request.

________________________________

In src/EnergyPlus/DataHeatBalance.cc<https://urldefense.com/v3/__https://github.com/NREL/EnergyPlus/pull/9639*discussion_r970128068__;Iw!!DZ3fjg!6ounRA5PEdmN4-vGIGxhOdD4hnkb9Uz6_WJKq3YO9IIik_jHSem9EhiQExRtCdzR3NVKjc8hgMDjlqZEtVQehLNE3No$>:

> +        0.0,       // None
+        0.1197548, // Wall
+        0.1620212, // Floor
+        0.1074271, // Roof
+        0.0,       // IntMass
+        0.0,       // Detached_B
+        0.0,       // Detached_F
+        0.1197584, // Window
+        0.1197584, // GlassDoor
+        0.1197584, // Door
+        0.0,       // Shading
+        0.0,       // Overhang
+        0.0,       // Fin
+        0.0,       // TDD_Dome
+        0.0        // TDD_Diffuser
+    };


try static_assert.

**Comment by Myoldmopar** [comment]
Caught all the way up on the conversation, this seems like it's ready.  Building locally for testing and reviewing the code now.

File: src/EnergyPlus/DataHeatBalance.cc
Comment (Myoldmopar) [inline src/EnergyPlus/DataHeatBalance.cc:1323]: This is fine, and as I mentioned on the call, if the _size_ of the SurfaceClass enum changes, the compiler should warn/error that this initializer list is the wrong size.  Now, if someone reorders the enums or something, then well, we'll have to rely on your comment there.  But that sort of thing goes for a lot of of stuff in E+.  This is good.

File: src/EnergyPlus/DataHeatBalance.cc
Comment (Myoldmopar) [inline src/EnergyPlus/DataHeatBalance.cc:1348]: :+1:

File: src/EnergyPlus/DataSurfaces.hh
Comment (Myoldmopar) [inline src/EnergyPlus/DataSurfaces.hh:113]: The comments here are a _little_ oddly placed.  I would have just:
```
enum class Foo {
  // comment 1
  // comment 2
  Invalid = -1,
  Bar,
  Num
}
```

Nothing to worry about for now, just noting.

File: tst/EnergyPlus/unit/DataHeatBalance.unit.cc
Comment (Myoldmopar) [reply tst/EnergyPlus/unit/DataHeatBalance.unit.cc:1364]: Agreed.

File: doc/output-details-and-examples/src/output-files/eplusout-eio.tex
Comment (Myoldmopar) [inline doc/output-details-and-examples/src/output-files/eplusout-eio.tex:1924]: I'm not going to request it be changed now, but in the future, please avoid raw quotation marks in the LaTeX source.  Use two backticks to start the quote and two apostrophes to end it.

**Review by Myoldmopar — APPROVED** [review]
Changes look good.

**Comment by Myoldmopar** [comment]
Everything runs happy locally as well.  Merging this, thanks @RKStrand and @rraustad

**Pull request merged**
